File: src/algorithm/template/layout/field/barcode_field.py
from src.algorithm.template.layout.field.base import Field, ScanBox
from src.algorithm.template.layout.field.field_drawing import BarcodeFieldDrawing
from src.utils.parsing import default_dump
from src.utils.shapes import ShapeUtils


class BarcodeField(Field):
    def __init__(
        self,
        direction,
        empty_value,
        field_block,
        field_detection_type,
        field_label,
        origin,
    ):
        super().__init__(
            direction,
            empty_value,
            field_block,
            field_detection_type,
            field_label,
            origin,
        )

    def get_drawing_instance(self):
        return BarcodeFieldDrawing(self)

    def setup_scan_boxes(self, field_block):
        scan_zone = field_block.scan_zone
        origin = field_block.origin
        field = self
        # TODO: support for multiple scan zones per field (grid structure)
        field_index = 0
        scan_box = BarcodeScanBox(field_index, field, origin, scan_zone)
        self.scan_boxes: list[BarcodeScanBox] = [scan_box]

    # Make the class serializable
    def to_json(self):
        return {
            key: default_dump(getattr(self, key))
            for key in [
                "field_label",
                "direction",
                "scan_boxes",
            ]
        }


class BarcodeScanBox(ScanBox):
    def __init__(self, field_index, field: BarcodeField, origin, scan_zone):
        dimensions = scan_zone["dimensions"]
        margins = scan_zone["margins"]
        super().__init__(field_index, field, origin, dimensions, margins)
        self.zone_description = {"origin": origin, "label": self.name, **scan_zone}
        # Compute once for reuse
        self.scan_zone_rectangle = ShapeUtils.compute_scan_zone_rectangle(
            self.zone_description, include_margins=True
        )

File: omrchecker-js/packages/core/src/processors/layout/field/barcodeField.ts
/**
 * Barcode field class (stub for future implementation).
 *
 * TypeScript port of src/processors/layout/field/barcode_field.py
 * Stub implementation - full port deferred to future phase.
 */

import { defaultDump } from '../../../utils/parsing';
import type { FieldBlock } from '../fieldBlock/base';
import { Field, ScanBox } from './base';
import { BarcodeFieldDrawing, type FieldDrawing } from './fieldDrawing';

/**
 * Barcode scan box (stub).
 */
export class BarcodeScanBox extends ScanBox {
  constructor(
    fieldIndex: number,
    field: BarcodeField,
    origin: [number, number],
    _scanZone: unknown
  ) {
    // Stub implementation
    super(fieldIndex, field, origin, [0, 0], { top: 0, bottom: 0, left: 0, right: 0 });
    // TODO: Implement barcode scan box setup
  }

  toJSON(): Record<string, unknown> {
    return {
      field_label: this.fieldLabel,
      name: this.name,
      x: this.x,
      y: this.y,
      origin: defaultDump(this.origin),
    };
  }
}

/**
 * Barcode field class (stub).
 */
export class BarcodeField extends Field {
  public scanZone?: unknown;

  constructor(
    direction: 'horizontal' | 'vertical',
    emptyValue: string,
    fieldBlock: FieldBlock,
    fieldDetectionType: 'BARCODE_QR',
    fieldLabel: string,
    origin: [number, number]
  ) {
    super(direction, emptyValue, fieldBlock, fieldDetectionType, fieldLabel, origin);
    this.scanZone = fieldBlock.scanZone;
  }

  setupScanBoxes(fieldBlock: FieldBlock): void {
    const scanZone = fieldBlock.scanZone;
    const origin = fieldBlock.origin;
    // TODO: support for multiple scan zones per field (grid structure)
    const fieldIndex = 0;
    const scanBox = new BarcodeScanBox(fieldIndex, this, origin, scanZone);
    this.scanBoxes = [scanBox];
  }

  getDrawingInstance(): FieldDrawing {
    return new BarcodeFieldDrawing(this);
  }

  toJSON(): Record<string, unknown> {
    return {
      field_label: this.fieldLabel,
      direction: this.direction,
      scan_boxes: this.scanBoxes.map((box) => box.toJSON()),
    };
  }
}

